docs/source/config.md
# Configuration for Evaluation/Verification

### Introduction

This page provides detailed documentation for configuring the NWM Evaluation Manager (nwm-eval-mgr) tool for a variety of simulation evaluation or forecast verification applications.

Template files, sample configuration files, and schemas for all configuration fields and subfields are included below. You can navigate to individual configuration files or schema sections using the tabs on the right or the Table of Contents below.

- `config_template.yaml`: Config template generated from the pydantic model, with default or example values defined for each field. This can be used as a starting point for creating your own configuration files.
- `config_ngencerf.yaml`: Sample config for verifying a single ngenCERF forecast at one location.
- `config_hindcast.yaml`: Sample config for verifying multiple ngenCERF hindcasts at one location.
- `config_nwm.yaml`: Sample config for verifying operational NWM v3 forecasts across multiple locations and domains using data retrieved from Google Cloud Storage (GCS).
- `config_ngensim.yaml`: Sample config for evaluating large-scale NGEN simulations (e.g., from regionalization) across multiple locations, VPUs, or NWM domains.

### Table of Contents

  - [Sample Files](#sample-files)
    - [`config_template.yaml`](#config-template-yaml)
    - [`config_ngencerf.yaml`](#config-ngencerf-yaml)
    - [`config_hindcast.yaml`](#config-hindcast-yaml)
    - [`config_nwm.yaml`](#config-nwm-yaml)
    - [`config_ngensim.yaml`](#config-ngensim-yaml)
  - [Schemas](#schemas)
    - [`BarChartConfig` (inherits from `BasePlotConfig`)](#barchartconfig-inherits-from-baseplotconfig)
    - [`BasePlotConfig` (inherits from `LeadTimesMixin`)](#baseplotconfig-inherits-from-leadtimesmixin)
    - [`BoxPlotConfig` (inherits from `BasePlotConfig`)](#boxplotconfig-inherits-from-baseplotconfig)
    - [`FilePathsConfig`](#filepathsconfig)
    - [`FlowObservationConfig`](#flowobservationconfig)
    - [`GeneralConfig`](#generalconfig)
    - [`HistogramConfig` (inherits from `BasePlotConfig`)](#histogramconfig-inherits-from-baseplotconfig)
    - [`LeadTimesMixin`](#leadtimesmixin)
    - [`LocationFilter`](#locationfilter)
    - [`MetricsConfig` (inherits from `LeadTimesMixin`)](#metricsconfig-inherits-from-leadtimesmixin)
    - [`NWMForecastConfig`](#nwmforecastconfig)
    - [`PairDataConfig`](#pairdataconfig)
    - [`PlotsConfig`](#plotsconfig)
    - [`ReferenceTimesMixin`](#referencetimesmixin)
    - [`SpatialMapConfig` (inherits from `BasePlotConfig`)](#spatialmapconfig-inherits-from-baseplotconfig)
    - [`TablePlotConfig` (inherits from `BasePlotConfig`)](#tableplotconfig-inherits-from-baseplotconfig)
    - [`TimeSeriesConfig` (inherits from `BasePlotConfig, ReferenceTimesMixin`)](#timeseriesconfig-inherits-from-baseplotconfig-referencetimesmixin)
    - [`USGSConfig`](#usgsconfig)

### Sample Files


(config-template-yaml)=
#### `config_template.yaml`

Config template generated from the pydantic model, with default or example values defined for each field. This can be used as a starting point for creating your own configuration files.

```yaml
general:     # General configuration for the evaluation, including dataset information and evaluation settings.
  steps:     # Steps to run in the evaluation process.
    fetch_fcst_data: True
    fetch_obs_data: True
    pair_data: True
    compute_metrics: True
    plot_metrics: True
  domain: 'conus'     # Domain for the evaluation. Valid options are 'conus', 'hi', 'ak', 'prvi' (case insensitive).
  assemble_domain: False     # Whether to assemble results across VPUs for the CONUS domain.
  location_set_name: 'usgs_01123000'     # Name for the set of locations to evaluate.
  location_list: ['01123000', '01123500']     # List of specific locations to include in the evaluation. If None, all locations in the crosswalk file will be used.
  location_type: 'usgs_gage'     # Type of locations to evaluate, which determines how locations are identified and processed.
  location_filter:     # Configuration for filtering locations based on column values.
    columns: ['vpu_id', 'status']
    values: ['03S', 'USGS-active']
  location_group_size: 200     # Number of locations to process in each group when pairing forecast and observation data.
  variable_name: 'streamflow'     # Name of the variable to evaluate. Currently only 'streamflow' is supported, but this field is included for future extensibility.
  nwm_configuration: 'ngen'     # NWM configuration to evaluate.
  dataset_name: ['noah_cfes', 'noah_topmodel']     # User-specified name(s) for the dataset(s) to evaluate.
  nwm_version: ['ngen', 'nwm30']     # NWM version(s) to evaluate.
  forecast_start_date: ['2022-12-01 00:00:00', '2022-12-15 00:00:00']     # Start date(s) for the forecast data to evaluate.
  forecast_end_date: ['2022-12-31 00:00:00', '2023-01-15 00:00:00']     # End date(s) for the forecast data to evaluate.
  eval_start_date: ['2022-12-11 00:00:00', '2022-12-25 00:00:00']     # Start date(s) for the evaluation period.
  eval_end_date: ['2022-12-31 00:00:00', '2023-01-15 00:00:00']     # End date(s) for the evaluation period.
  separate_calibrated: False     # Whether to distinguish calibrated and regionalized locations in the evaluation
  log_level: 'info'     # Logging level. Valid options (case insensitive): debug, info, warning, error, critical, severe, fatal
file_paths:     # Configuration for file paths used in the evaluation, including input data and output directories.
  base_dir: '~/ngen_evaluation/'     # Root directory for storing data and outputs.
  location_list_file: ~/location_list.csv     # Path to a file containing the list of locations to evaluate
  crosswalk_file: ~/crosswalk.csv     # Path to a crosswalk file or a dictionary of crosswalk files.
  fcst_config_file: 'data/inputs/nwm_forecast_configuration.yaml'     # Path to the forecast configuration file defining parameters for different NWM configurations.
  fcst_data_file: '01123000_output.csv'     # Path to the forecast data file or a dictionary of forecast data files.
  fcst_data_dir: 'data/inputs/hindcasts/'     # Path to the directory containing forecast data files or a dictionary of directories.
  obs_data_file: 'data/inputs/obs/01123000_hourly_discharge.csv'     # Path to the observation data file. Both obs_data_file and obs_data_dir can be specified. The run will read all .csv and .parquet files in obs_data_dir and obs_data_file and remove duplicates.If neither obs_data_file nor obs_data_dir is provided, an observation data source must be specified in the 'flow_observation.usgs' section.
  obs_data_dir: 'data/inputs/obs/'     # Path to the observation data directory where one or more observation data files are stored. The run will look for all .csv and .parquet files in this directory. Each file can contain observation data for a single location (with the filename starting with the location identifier), or multiple locations with a 'location_id' column specifying the location identifiers.
  calib_param_file: 'data/inputs/calib_params.csv'     # Path to the calibration parameter file.
  txdot_gage_file: 'data/inputs/gage_files/tx_gauges.csv'     # Path to the TxDOT gage file.
  output_dir: 'ngen_evaluation/outputs/usgs_01123000/'     # Directory to save outputs such as paired data, computed metrics, and plots. 
  log_file: 'outputs/usgs_01123000/verification.log'     # Path to log file. Default: verification.log in {output_dir}
nwm_forecast:     # Configuration for NWM forecast data.
  data_source: 'ngenCERF'     # Data source for the NWM forecast. Valid options include 'ngenCERF', 'ngenSIM', 'hindcast', and 'GCS'.
  fetch_fcst: [True]     # Whether to fetch forecast data for each dataset, with local file existence check.
  output_type: 'channel_rt'     # Type of NWM output to retrieve. Currently only 'channel_rt' is supported. Only applicable when data_source='GCS'. 
  t_minus: [0, 1, 2]     # T-minus hours for which to retrieve NWM forecasts, only applicable for AnA runs on GCS.
  kerchunk_method: 'local'     # Preference for creating Kerchunk reference json files, only applicable for GCS data source.
  process_by_z_hour: True     # Whether to process NWM files by z-hour per day, only applicable for GCS data source.
  stepsize: 100     # Number of files to process in memory at once when processing GCS data without z-hour chunking.
  ignore_missing_file: True     # Whether to ignore missing NWM files and continue processing, only applicable for GCS data source.
  overwrite_output: False     # Whether to overwrite existing forecast data files.
  memory_per_worker_gb: 3     # Configurable memory (in GB) assigned to each worker or process.
flow_observation:     # Configuration for flow observation data.
  usgs: ''     # Configuration for USGS flow observations. If omitted, obs_data_file and/or obs_data_dir must be provided in file_paths section, 
pair_data:     # Configuration for paired data.
  overwrite: True     # Whether to overwrite existing paired data files.
  group_size: 200     # Number of locations to process in each group when pairing forecast and observation data.
metrics:     # Configuration for metrics.
  lead_times: ['all', '1-5', '5-10', '10-15', 'all_aggregated']     # Lead times to include in the evaluation.
  overwrite: True     # Whether to overwrite existing metric files. 
  library: 'nwm.eval'     # Library to use for metric computation. Valid options: nwm.eval, teehr. 
  metric_subset: 'all'     # Subset of metrics to compute, either 'all' or a list of metric names.
  metric_exclude: ['HSEG_FDC', 'MSEG_FDC', 'LSEG_FDC']     # List of metric names to exclude from metric_subset for computation. 
  threshold_categorical:     # Threshold for categorical metrics. Valid options for type are 'quantile' and 'absolute'. If 'quantile', the threshold will be determined as the specified quantile of the observed flow values. If 'absolute', the threshold will be the specified absolute flow value.
    value: 0.9
    type: quantile
  threshold_event:     # Threshold for event-based metrics. Valid options for type are 'quantile' and 'absolute'. If 'quantile', the threshold will be determined as the specified quantile of the observed flow values. If 'absolute', the threshold will be the specified absolute flow value.
    value: 0.9
    type: quantile
  file_format: 'parquet'     # File format for output files. Valid options are 'parquet' and 'csv'.
plots:     # Configuration for plots.
  histogram:     # Configuration for histogram plots.
    lead_times: ['all', '1-5', '5-10', '10-15', 'all_aggregated']     # Lead times to include in the evaluation.
    plot: False     # Whether to generate this type of plot.
    metric_subset: ['NSE', 'KGE']     # List of metric names to include in the plots.
    tag: ''     # Optional tag to include in plot titles and filenames.
    binning:     # Dictionary specifying the binning for histogram plots, with metric names as keys and lists of bin edges as values.
      NSE: [-1, -0.5, 0, 0.5, 1]
      KGE: [-1, -0.5, 0, 0.5, 1]
  boxplot:     # Configuration for box plot.
    lead_times: ['all', '1-5', '5-10', '10-15', 'all_aggregated']     # Lead times to include in the evaluation.
    plot: False     # Whether to generate this type of plot.
    metric_subset: ['NSE', 'KGE']     # List of metric names to include in the plots.
    tag: ''     # Optional tag to include in plot titles and filenames.
    show_outliers: False     # Whether to show outliers in box plots. If True, outliers will be shown as individual points. If False, outliers will be omitted.
  spatial_map:     # Configuration for spatial map plots.
    lead_times: ['all', '1-5', '5-10', '10-15', 'all_aggregated']     # Lead times to include in the evaluation.
    plot: False     # Whether to generate this type of plot.
    metric_subset: ['NSE', 'KGE']     # List of metric names to include in the plots.
    tag: ''     # Optional tag to include in plot titles and filenames.
    scaling:     # Dictionary specifying the scaling for spatial maps, with metric names as keys and lists of scaling ranges as values.
      NSE: [-0.5, 1.0]
      KGE: [-0.5, 1.0]
  time_series:     # Configuration for time series plots.
    reference_times: ['2022-12-01 00:00:00', '2022-12-15 00:00:00']     # Reference times (T0s) to include in the evaluation.
    lead_times: [1, 2, 3]     # List of lead times for time series plots. Each lead time should be a number representing the forecast lead time in hours.
    plot: False     # Whether to generate this type of plot.
    metric_subset: ['NSE', 'KGE']     # List of metric names to include in the plots.
    tag: ''     # Optional tag to include in plot titles and filenames.
  metric_table:     # Configuration for metric table plots.
    lead_times: ['all', '1-5', '5-10', '10-15', 'all_aggregated']     # Lead times to include in the evaluation.
    plot: False     # Whether to generate this type of plot.
    metric_subset: ['NSE', 'KGE']     # List of metric names to include in the plots.
    tag: ''     # Optional tag to include in plot titles and filenames.
  barchart:     # Configuration for bar chart plots.
    lead_times: ['all', '1-5', '5-10', '10-15', 'all_aggregated']     # Lead times to include in the evaluation.
    plot: False     # Whether to generate this type of plot.
    metric_subset: ['NSE', 'KGE']     # List of metric names to include in the plots.
    tag: ''     # Optional tag to include in plot titles and filenames.
```

(config-ngencerf-yaml)=
#### `config_ngencerf.yaml`

Sample config for verifying a single ngenCERF forecast at one location.

```yaml
general:   
  steps: # define which of the 5 steps to run; each step can be run independently,
    fetch_fcst_data: true
    fetch_obs_data: true
    pair_data: true
    compute_metrics: true
    plot_metrics: true
  location_set_name: usgs_01123000  # user-specified name for the set of locations
  location_list: ['01123000'] # list of usgs gage IDs
  location_type: usgs_gage # location type
  variable_name: streamflow # currently only 'streamflow' is supported
  nwm_configuration: short_range # must match one of the configurations defined in fcst_config_file  
  dataset_name: [noah_cfes] # user-specified name of datasets (e.g., formualtion name
  nwm_version: [ngen] # list of NWM versions; must have the same length as 'dataset_name'
  forecast_start_date: ['2022-12-01 00:00:00'] # list of start dates for verification; must be same length as 'dataset_name'
  forecast_end_date: ['2022-12-01 00:00:00'] # list of end dates for verification; must be same length as 'dataset_name'
  log_level: INFO # logging level (case insensitive); options are DEBUG, INFO, WARNING, SEVERE, FATAL, CRITICAL, ERROR

file_paths:
  base_dir: data/ # root directory to store data/outputs for verification
  crosswalk_file: '{base_dir}/inputs/gage_files/usgs_{nwm_version}_crosswalk_all_domains.parquet' # crosswalk file mapping gage IDs to NWM/ngen link IDs
  fcst_config_file: '{base_dir}/inputs/nwm_forecast_configuration.yaml' # NWM forecast configuration file
  fcst_data_file: '{base_dir}/inputs/{location_set_name}/{dataset_name}/{nwm_configuration}.csv' # forecast data directory (must be specified when nwm_forecast.data_source is set to ngenCERF)
  output_dir: '{base_dir}/outputs/{location_set_name}' # output directory to store all datasets, metrics and plots
  log_file: '{output_dir}/verification.log'

nwm_forecast:
  data_source: ngenCERF    # Specifies the source to retrieve the model forecast or simulation data. Currently supported options: GCS, ngenCERF, ngenSIM

flow_observation:
  usgs: # flow obs retrieved will be stored at [data_dir_root]/[location_set_name]/usgs
    chunk_by: month   # data are saved in parquet files by day/month/year
    overwrite_output: True  # If True, existing output files will be overwritten; if False, existing files are retained
    memory_per_worker_gb: 3  # configurable memory (in GB) assigned to each worker

pair_data:  
  overwrite: true # whether to overwrite existing joined parquet files
  group_size: 200 # break large location list into smaller groups (for pair_data and cacl_metrics steps) to avoid potential memory issues and to speed up processing

metrics:  
  overwrite: true # whether to overwrite existing metric files
  library: nwm.eval # currently supported options: teehr, nwm.eval
  metric_subset: 'all'
  threshold_categorical:
    value: 0.9 # threshold value to be used for categorical metrics in nwm.eval 
    type: quantile # type of threshold for categorical metrics in nwm.eval; options are 'quantile' or 'absolute'
  threshold_event:
    value: 0.9 # threshold value to be used for event-based metrics in nwm.eval
    type: quantile # type of threshold for event-based metrics in nwm.eval; options are 'quantile' or 'absolute'
  lead_times: ['all_aggregated'] # list of lead times (in hours) for which metrics should be calculated for  
  file_format: parquet # file format for metrics. Options are: parquet, csv. Default is parquet.

plots:
  time_series:
    plot: true  # whether to create time series plots
  metric_table:
    plot: true  # whether to create metric table plots
  barchart:
    plot: true  # whether to create barchart plots
```

(config-hindcast-yaml)=
#### `config_hindcast.yaml`

Sample config for verifying multiple ngenCERF hindcasts at one location.

```yaml
general:   
  steps: # define which of the 5 steps to run; each step can be run independently,
    fetch_fcst_data: true
    fetch_obs_data: true
    pair_data: true
    compute_metrics: true
    plot_metrics: true
  location_set_name: 'usgs_01123000'  # user-specified name for the set of locations
  location_list: ['01123000'] # list of usgs gage IDs
  location_type: usgs_gage # location type
  variable_name: streamflow # currently only 'streamflow' is supported
  nwm_configuration: short_range  # must match one of the configurations defined in fcst_config_file  
  dataset_name: [noah_cfes, noah_sac] # user-specified name of datasets (e.g., formualtion name)
  nwm_version: [ngen, ngen] # list of NWM versions; must have the same length as 'dataset_name'
  forecast_start_date: ['2025-08-20 00:00:00', '2025-08-20 00:00:00'] # list of start dates for verification; must be same length as 'dataset_name'
  forecast_end_date: ['2025-08-22 06:00:00', '2025-08-22 06:00:00'] # list of end dates for verification; must be same length as 'dataset_name'
  log_level: INFO # logging level (case insensitive); options are DEBUG, INFO, WARNING, SEVERE, FATAL, CRITICAL, ERROR

file_paths:
  base_dir: data/ # root directory to store data/outputs for verification
  crosswalk_file: '{base_dir}/inputs/gage_files/usgs_{nwm_version}_crosswalk_all_domains.parquet' # crosswalk file mapping gage IDs to NWM/ngen link IDs
  fcst_config_file: '{base_dir}/inputs/nwm_forecast_configuration.yaml' # NWM forecast configuration file
  fcst_data_dir: '{base_dir}/inputs/{location_set_name}/{dataset_name}/hind_run1' # forecast data directory or file (must be specified when nwm_forecast.data_source is set to ngenCERF or hinscast)
  fcst_data_file: '01123000_output.csv' # forecast data file path pattern (must be specified when nwm_forecast.data_source is set to ngenCERF or hinscast)
  output_dir: '{base_dir}/outputs/{location_set_name}_hindcast' # output directory to store all datasets, metrics and plots
  log_file: '{output_dir}/verification.log'

nwm_forecast:
  data_source: hindcast    # Specifies the source to retrieve the model forecast or simulation data. Currently supported options: GCS, ngenCERF, ngenSIM, hindcast

flow_observation:
  usgs: # flow obs retrieved will be stored at [data_dir_root]/[location_set_name]/usgs
    chunk_by: month   # data are saved in parquet files by day/month/year
    overwrite_output: True  # If True, existing output files will be overwritten; if False, existing files are retained
    memory_per_worker_gb: 3  # configurable memory (in GB) assigned to each worker

pair_data:  
  overwrite: true # whether to overwrite existing joined parquet files
  group_size: 200 # break large location list into smaller groups (for pair_data and cacl_metrics steps) to avoid potential memory issues and to speed up processing

metrics:  
  overwrite: true # whether to overwrite existing metric files
  library: nwm.eval # currently supported options: teehr, nwm.eval
  metric_subset: 'all'
  threshold_categorical:
    value: 0.9 # threshold value to be used for categorical metrics in nwm.eval 
    type: quantile # type of threshold for categorical metrics in nwm.eval; options are 'quantile' or 'absolute'
  threshold_event:
    value: 0.9 # threshold value to be used for event-based metrics in nwm.eval
    type: quantile # type of threshold for event-based metrics in nwm.eval; options are 'quantile' or 'absolute'
  lead_times: [all, 1-5, 6-10, 11-18, all_aggregated] # list of lead times (in hours) for which metrics should be calculated for  
  file_format: parquet # file format for metrics. Options are: parquet, csv. Default is parquet.

plots:
  time_series:
    plot: true  # whether to create time series plots
    lead_times: [1, 6, 12, 18] # list of lead times (in hours) for which time series plots should be created
    reference_times: ['2025-08-20 00:00:00', '2025-08-21 00:00:00', '2025-08-22 00:00:00'] # list of forecast reference times for which time series plots should be created
  metric_table:
    plot: true  # whether to create metric table plots
    lead_times: [1, 10, 1-5, all_aggregated] # list of lead times (in hours) for which metric table plots should be created
  barchart:
    plot: true  # whether to create barchart plots
    metric_subset: [KGE, NSE, CORR, NNSE, PBIAS]
    lead_times: [1, 5, 10, 18, 1-5, 6-10, 11-18, "all_aggregated"] # list of lead times (in hours) for which barchart plots should be created
```

(config-nwm-yaml)=
#### `config_nwm.yaml`

Sample config for verifying operational NWM v3 forecasts across multiple locations and domains using data retrieved from Google Cloud Storage (GCS).

```yaml
general: # define which of the 5 steps to run; each step can be run independently, assuming data from previous steps (from previous runs) are available for use
  steps:
    fetch_fcst_data: true
    fetch_obs_data: true
    pair_data: true
    compute_metrics: true
    plot_metrics: true

  location_set_name: calib_basin_group1 # user-specified name for the set of locations
  variable_name: streamflow # currently only 'streamflow' is supported
  nwm_configuration: short_range   # Currently, only short_range, short_range_alaska, short_range_hawaii, and short_range_puertorico are supported
  dataset_name: [v3_sep, v3_oct] # user-specified name of datasets (used for creating sub-folders for datasets and plots) 
  nwm_version: [nwm30, nwm30] # list of NWM versions; must have the same length as 'dataset_name'.
  forecast_start_date: ['2024-09-01 00:00:00', '2024-10-01 00:00:00'] # list of start date for forecast verification period; list must have the same length as 'dataset_name'
  forecast_end_date: ['2024-09-30 23:00:00', '2024-10-30 23:00:00'] # list of end date for forecast verification period; list must have the same length as 'dataset_name'
  log_level: INFO # logging level (case insensitive); options are DEBUG, INFO, WARNING, SEVERE, FATAL, CRITICAL, ERROR

file_paths:  
  base_dir: ~/repos/nwm-eval-mgr/data/  # root directory to store the downloaded NWM forecast and flow observation data
  location_list_file: "{base_dir}/inputs/gage_files/usgs_gages_link_CONUS_calib100.csv" # [optional] file path for the list of locations (e.g., gage IDs or NWM link IDs)
  crosswalk_file: '{base_dir}/inputs/gage_files/usgs_{nwm_version}_crosswalk_all_domains.parquet' # parquet files containing the crosswalk between NWM feature_id and usgs gage id  
  fcst_config_file: '{base_dir}/inputs/nwm_forecast_configuration.yaml' # NWM forecast configuration file  
  output_dir: '{base_dir}/outputs/{location_set_name}' # output directory to store all datasets and plots
  log_file: '{output_dir}/verification.log'

nwm_forecast:
  data_source: GCS    # Specifies the source to retrieve the model forecast or simulation data. Currently supported options: GCS, ngenCERF, ngenSIM
  fetch_fcst: [true, true]  # whether to fetch the forecast data for each dataset listed in ['general']['dataset_name']  
  output_type: channel_rt
  t_minus: [0, 1, 2]  # Only used if an assimilation run is selected
  kerchunk_method: local  
  process_by_z_hour: true  
  stepsize: 100  # Only used if PROCESS_BY_Z_HOUR = False. Controls how many files are processed in memory at once
  ignore_missing_file: false  # If True, the missing file(s) will be skipped and the process will resume; if False, TEEHR will fail if a missing NWM file is encountered

nwm_eval/nwm_eval/settings.py
"""Settings and configurations for NWM evaluation.

Functions:
    - `get_metric_colormap`: Define the color maps for each metric for creating the spatial maps (in create_plots.py).
    - `get_metric_bins`: Define the binnings for each metric for creating the histograms (in create_plots.py).
    - `data_paths`: Define the data paths for observations, forecasts, and outputs based on the configuration file.

Constants:
    - `metric_value_bins_default`: Default bins for each metric for creating the histograms (in create_plots.py).
    - `metric_color_scale_default`: Default color maps for each metric for creating the spatial maps (in create_plots.py).
    - `metric_groups`: Define different groups of metrics based on whether higher or lower values are better, and whether absolute values are applicable.
    - `dict_teehr_metrics`: Mapping of metric names to TEEHR metric names.
    - `dict_nwm_eval_metrics`: Mapping of metric names to NWM evaluation metric names.
    - `conus_vpu_list`: List of CONUS VPU identifiers.
    - `default_txdot_gage_list`: List of default TXDOT gage identifiers.

"""

import logging
from pathlib import Path

logger = logging.getLogger(__name__)

# bins for each metric for creating the histograms (in create_plots.py)
metric_value_bins_default = {
    "KGE": [float("-inf"), -1, -0.5, 0, 0.2, 0.4, 0.6, 0.8, 1.0],
    "NSE": [float("-inf"), -1, -0.5, 0, 0.2, 0.4, 0.6, 0.8, 1.0],
    "NNSE": [0, 0.2, 0.4, 0.5, 0.6, 0.7, 0.8, 1.0],
    "CORR": [-1, -0.5, 0, 0.2, 0.4, 0.6, 0.8, 1.0],
}

# color maps for each metric for creating the spatial maps (in create_plots.py)
metric_color_scale_default = dict(
    KGE=(-0.5, 1),
    NSE=(-0.5, 1),
    CORR=(-0.5, 1),
    NNSE=(0, 1),
)

# define different groups of metrics
metric_groups = dict(
    higher_is_better=[
        "CORR",
        "sCORR",
        "NSE",
        "NNSE",
        "NSElog",
        "NSEwt",
        "KGE",
        "KGE1",
        "KGE2",
        "R2",
        "POD",
        "CSI",
        "n_obs",
        "n_mod",
        "min_obs",
        "min_mod",
        "max_obs",
        "max_mod",
        "mean_obs",
        "mean_mod",
        "sum_obs",
        "sum_mod",
        # "pt_obs",
        # "pt_mod",
    ],
    lower_is_better=[
        "ME",
        "MAE",
        "MSE",
        "RMSE",
        "RMAE",
        "PBIAS",
        "RBIAS",
        "MBAIS",
        "aprBIAS",
        "RSR",
        "HSEG_FDC",
        "MSEG_FDC",
        "LSEG_FDC",
        "FAR",
        "FBIAS",
        "PKBIAS",
        "PKTE",
        "EVBIAS",
        "var_obs",
        "var_mod",
        "max_delta",
        # "pt_err",
    ],
    abs_applicable=["ME", "PBIAS", "HSEG_FDC", "MSEG_FDC", "LSEG_FDC"],
)


def get_metric_colormap(conf: dict, plot_type: str) -> dict:
    """Define the color maps for each metric for creating the spatial maps (in create_plots.py)."""
    metric_cmaps = dict()
    for m1 in conf["metric_subset"]:
        metric_cmaps[m1] = dict()

        # define scale
        if (
            "scaling" in conf.keys()
            and m1 in conf["scaling"]
            and conf["scaling"][m1] is not None
        ):
            metric_cmaps[m1]["clim"] = tuple(conf["scaling"][m1])
        elif m1 in metric_color_scale_default.keys():
            metric_cmaps[m1]["clim"] = metric_color_scale_default[m1]
        else:
            logger.info(
                f"scaling not defined for {m1}; metric data will not be scaled when creating {plot_type}"
            )
            metric_cmaps[m1]["clim"] = (float("nan"), float("nan"))

        # define color maps
        if plot_type == "map":
            if m1 in metric_groups["higher_is_better"]:
                metric_cmaps[m1]["cmap"] = "rainbow_r"  # cc.rainbow[::-1]
            elif m1 in metric_groups["lower_is_better"]:
                metric_cmaps[m1]["cmap"] = "rainbow"  # cc.rainbow
            else:
                logger.info(
                    f"metric orientation not defined for {m1}; set to default (rainbow_r)"
                )

    return metric_cmaps


def get_metric_bins(conf: dict) -> dict:
    """Define the binnings for each metric for creating the histograms (in create_plots.py)."""
    metric_bins = dict()
    for m1 in conf["metric_subset"]:
        if (
            "binning" in conf.keys()
            and m1 in conf["binning"]
            and conf["binning"][m1] is not None
        ):
            bins = conf["binning"][m1]
            bins = [float(x.lower()) if type(x) is str else x for x in bins]
            metric_bins[m1] = bins
        elif m1 in metric_value_bins_default.keys():
            metric_bins[m1] = metric_value_bins_default[m1]
        else:
            metric_bins[m1] = []
            logger.info(f"binning is not defined for {m1}; use equal-width bins")

    return metric_bins


dict_teehr_metrics = {
    "n_obs": "primary_count",
    "n_mod": "secondary_count",
    "min_obs": "primary_minimum",
    "min_mod": "secondary_minimum",
    "max_obs": "primary_maximum",
    "max_mod": "secondary_maximum",
    "mean_obs": "primary_average",
    "mean_mod": "secondary_average",
    "sum_obs": "primary_sum",
    "sum_mod": "secondary_sum",
    "var_obs": "primary_variance",
    "var_mod": "secondary_variance",
    "max_delta": "max_value_delta",
    "NSE": "nash_sutcliffe_efficiency",
    "NNSE": "nash_sutcliffe_efficiency_normalized",
    "KGE": "kling_gupta_efficiency",
    "KGE1": "kling_gupta_efficiency_mod1",
    "KGE2": "kling_gupta_efficiency_mod2",
    "ME": "mean_error",
    "MAE": "mean_absolute_error",
    "MSE": "mean_squared_error",
    "RMSE": "root_mean_squared_error",
    # "pt_obs": "primary_max_value_time",  # these timestamps based metrics are currently not supported
    # "pt_mod": "secondary_max_value_time",
    # "pt_err": "max_value_timedelta",
    "RBIAS": "relative_bias",
    "MBAIS": "multiplicative_bias",
    "RMAE": "mean_absolute_relative_error",
    "CORR": "pearson_correlation",
    "sCORR": "spearman_correlation",
    "R2": "r_squared",
    "aprBIAS": "annual_peak_relative_bias",
}

dict_nwm_eval_metrics = {
    "CORR": "pearson correlation",
    "NSE": "nash_sutcliffe_efficiency",
    "NNSE": "nash_sutcliffe_efficiency_normalized",
    "NSElog": "logrithmic nash_sutcliffe_efficiency",
    "NSEwt": "weighted NSE and NSElog",
    "KGE": "kling_gupta_efficiency",
    "MAE": "mean_absolute_error",
    "RMSE": "root_mean_squared_error",
    "PBIAS": "percent_bias",
    "RSR": "RMSE_observation_std_ratio",
    "HSEG_FDC": "pbias_high_flow_FDC",
    "MSEG_FDC": "pbias_medium_flow_FDC",
    "LSEG_FDC": "pbias_low_flow_FDC",
    "POD": "probability_of_detection",
    "FAR": "false_alarm_ratio",
    "CSI": "critical_success_index",
    "FBIAS": "frequency_bias",
    "PKBIAS": "percent_peak_flow_bias",
    "PKTE": "peak_timing_error",
    "EVBIAS": "event_volume_bias",
}


def data_paths(conf: dict) -> dict:
    """Define the data paths for observations, forecasts, and outputs based on the configuration file."""
    conf1 = conf["general"]
    conf2 = conf["file_paths"]
    out_dir = Path(conf2["output_dir"])
    config = conf1["nwm_configuration"]

    # paths for all observations
    obs_dir = out_dir / "usgs"

    # paths for forecast datasets
    fcst_data_dir = {}
    fcst_json_dir = {}
    fcst_data_link_dir = {}
    paired_data_file = {}
    metric_file = {}
    for idx, dataset in enumerate(conf1["dataset_name"]):
        nwm_ver = conf1["nwm_version"][idx]

        # create output directories based on NWM version
        fcst_json_dir[dataset] = out_dir / nwm_ver / "zarr" / config
        fcst_data_dir[dataset] = out_dir / nwm_ver / "timeseries" / config

        # create additional directory for storing symbolic links to parquet files required for each dataset
        fcst_data_link_dir[dataset] = (
            out_dir / conf1["dataset_name"][idx] / conf1["nwm_configuration"]
        )

        # path for joined parquet files (note in pair_data.py, 'group*' will be added to the file name for individual location groups)
        filename = f"{dataset}.{nwm_ver}.{conf1['nwm_configuration']}.joined.parquet"
        paired_data_file[dataset] = out_dir / "joined" / filename

        # path for metric output files
        metric_file[dataset] = (
            out_dir / "metrics" / filename.replace("joined", "metrics")
        )

        if conf["metrics"]["file_format"] == "csv":
            metric_file[dataset] = metric_file[dataset].with_suffix(".csv")

    # path for plots
    plot_dir = out_dir / "plots" / conf1["nwm_configuration"]

    # path for crosswalk file
    cwt_file = {}
    if conf2.get("crosswalk_file"):
        for ver1 in list(set(conf1["nwm_version"])):
            if ver1 in conf2["crosswalk_file"]:
                cwt_file[ver1] = Path(conf2["crosswalk_file"][ver1])
            else:
                msg = (
                    f"crosswalk file not found for {ver1}; "
                    f"please check if crosswalk_file is correctly defined in the config file"
                )
                logger.error(msg)
                raise ValueError(msg)

    # assemble all paths into a dictionary
    data_paths = {
        "fcst": fcst_data_dir,
        "fcst_json": fcst_json_dir,
        "fcst_link": fcst_data_link_dir,
        "obs": obs_dir,
        "joined": paired_data_file,
        "metrics": metric_file,
        "plots": plot_dir,
        "crosswalk": cwt_file,
    }

    return data_paths


conus_vpu_list = [
    "01",
    "02",
    "03N",
    "03S",
    "03W",
    "04",
    "05",
    "06",
    "07",
    "08",
    "09",
    "10U",
    "10L",
    "11",
    "12",
    "13",
    "14",
    "15",
    "16",
    "17",
    "18",
]

default_txdot_gage_list = [
    "08030530",
    "08031005",
    "08031020",
    "08041788",
    "08041790",
    "08041940",
    "08041945",
    "08041970",
    "08042455",
    "08042468",
    "08042470",
    "08042515",
    "08042539",
    "08064990",
    "08065080",
    "08065310",
    "08065340",
    "08065420",
    "08065700",
    "08065820",
    "08065925",
    "08066087",
    "08066138",
    "08066380",
    "08067280",
    "08067505",
    "08067520",
    "08067653",
    "08068020",
    "08068025",
    "08070220",
    "08070550",
    "08070900",
    "08076990",
    "08077110",
    "08077640",
    "08077670",
    "08077888",
    "08078400",
    "08078890",
    "08078910",
    "08078935",
    "08097000",
    "08098295",
    "08100950",
    "08102730",
    "08108705",
    "08108710",
    "08109310",
    "08110520",
    "08111006",
    "08111051",
    "08111056",
    "08111070",
    "08111080",
    "08111085",
    "08111090",
    "08111110",
    "08117375",
    "08117403",
    "08117857",
    "08117858",
    "08162580",
    "08163720",
    "08163880",
    "08163900",
    "08164150",
    "08164200",
    "08164410",
    "08167000",
    "08169778",
    "08173210",
    "08174545",
    "08180990",
    "08189298",
    "08189320",
    "08189520",
    "08189585",
    "08189590",
]
